Ac 20 created fast api config files. Able to download from S3

Omermigdal opened the pull request:
Notice that what i figured out is that in order to be able to connect to S3 you have to configure your AWS credentials as enviroment variables.

Omermigdal committed:
created empty FastAPI config files

Omermigdal committed:
added poetry files

Omermigdal committed:
added FastAPI configuration

Omermigdal committed:
linted

Omermigdal committed:
WOP
trying to reach some acces to s3

Omermigdal committed:
Updated the enviroment variables to contain AWS credentials.
Succeded downloading from S3 .

Omermigdal committed:
Updated the enviroment variables to contain AWS credentials.
Succeded downloading from S3 .

Omermigdal committed:
Merge pull request #11 from mta-final-project/create_FastAPI_config_files

Create fast api config files

tal-alon commented on source/FilesServer/pyproject.toml:
```diff
@@ -0,0 +1,21 @@
+uvicorn = "^0.29.0"
+pydantic = "^2.6.4"
+pydantic-settings = "^2.2.1"
+ruff = "^0.3.4"
```
ruff should be a dev dependency so we wouldn't unnecessarily install in in production.
remove it with "poetry remove ruff"
and reinstall with "poetry add -G dev ruff"

Omermigdal replied:
done

tal-alon commented on source/FilesServer/app/mongo.py:
if we are not using mongo for additional metadata please delete this file (and any other mongo related code)

Omermigdal replied:
done

tal-alon commented on source/FilesServer/app/settings.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class MongoSettings(BaseModel):
+    url: str = "mongodb://user:pass@localhost:27017"
```
same, not needed anymore

Omermigdal replied:
done

tal-alon commented on source/FilesServer/app/lifespan.py:
not needed since we don't need to connect to mongo before the server can start

Omermigdal replied:
done

tal-alon commented on source/FilesServer/app/api.py:
```diff
@@ -0,0 +1,72 @@
+    region_name=os.getenv("AWS_REGION"),
+)
+
+AWS_BUCKET = "academease"
```
move to the settings class

tal-alon commented on source/FilesServer/app/api.py:
```diff
@@ -0,0 +1,72 @@
+
+# Check if the environment variables are set
+if (
+    not os.getenv("AWS_ACCESS_KEY_ID")
```
use pydantic settgins for this. i this that if you don't specify a default value pydantic will throw an error if some of the values weren't provided an env var (need to be verified, i'm not sure)

tal-alon commented on source/FilesServer/app/api.py:
please split this file into an endpoints.py file (just for the routes) and a service.py file for the boto3 logic

Omermigdal committed:
fixed all the comments:
- splitted api.py to endpoints and service
- moved aws credentials validation to pydantic
- deleted mongo related files

Omermigdal committed:
fixed s3_download endpoint using the SessionSettings..

tal-alon commented on source/FilesServer/app/endpoints.py:
```diff
@@ -0,0 +1,14 @@
+
+
+@router.get("/download", status_code=status.HTTP_200_OK)
+async def download(file_name: str | None = None):
```
add type annotations for the return value

tal-alon commented on source/FilesServer/app/endpoints.py:
```diff
@@ -0,0 +1,14 @@
+
+
+@router.post("/upload")
+async def upload(file: UploadFile | None = None):
```
same

tal-alon commented on source/FilesServer/app/service.py:
```diff
@@ -0,0 +1,26 @@
+        )
+    safe_file_name = quote(file_name)
+
+    contents = bucket.Object(key=file_name).get()["Body"].read()
```
we don't want to download the file on each request, it's a waste of aws operations (which we are charged for), network, ram and other computing resources. i think we want to generate (and return) a presigned url, which can be used by whoever called the api to download the file themselves.
read about it here:
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html
https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html

tal-alon commented on source/FilesServer/app/settings.py:
```diff
@@ -0,0 +1,39 @@
+    bucket_name: str = "academease"
+
+    @property
+    def bucket_instance(self):
```
i don't think this property belongs in the settings. the settings should contain only configurable, static values. the bucket_name for instance, but not the bucket_instance object

tal-alon commented on source/FilesServer/app/settings.py:
```diff
@@ -0,0 +1,39 @@
+
+
+class AWSSettings(BaseModel):
+    access_key_id: str = os.getenv("AWS_ACCESS_KEY_ID")
```
since this is a pydantic settings object you don't actually need to explicitly call os.getenv. this is essentially what pydantic will do anyway. However because of the "__" delimiter and env var name for the access_key_id field will be "AWS__ACCESS_KEY_ID". if you want it to have only "_" maybe you can change the env_nested_delimiter

Omermigdal committed:
WOP
trying to configure aws credentials in .env

Omermigdal committed:
moved bucket creation to service

Omermigdal committed:
reformatted code

Omermigdal committed:
added todo

Omermigdal committed:
untracking .env
added README.md

tal-alon reviewed (approved):
please open a new ticket for the remaining comment:
we don't want to download the file on each request, it's a waste of aws operations (which we are charged for), network, ram and other computing resources. i think we want to generate (and return) a presigned url, which can be used by whoever called the api to download the file themselves.
read about it here:
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html
https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html

Omermigdal commented:
I did. Its called "Modify code to use PresignedURL"
